## Zadanie 1.
1.  Wejdź  na  stronę NCBI (http://www.ncbi.nlm.nih.gov) lub Entrez (https://www.ncbi.nlm.nih.gov/gquery/) – pierwszy link zaprowadzi Cie na główną stronę NCBI, gdzie możesz się zapoznać z dostępnymi narzędziami; drugi wyświetla listę baz danych indeksowanych przez Entrez.
2. W pole wyszukiwania na górze strony wpisz: hemoglobin human - pokażą się rezultaty we wszystkich  bazach  NCBI  dla  tego  zapytania (w  ilu  bazach  danych  zostały  znalezione informacje?).
3. Wybierz bazę Gene,  a  w  wynikach pierwszy  znaleziony  rekord  o  nazwie  HBB - pojawia się  strona  podsumowująca  informacje  o  tym  genie.  Czego  możemy  się  dowiedzieć  o hemoglobinie w kolejnych sekcjach?
4. Znajdź sekcję NCBI Reference Sequences (RefSeq). Co wyróżnia bazę RefSeq?
5. Pobierz sekwencję genomową formacie FASTA w podsekcji Genomic.
6. W podsekcji mRNA and Protein przejdź do linka oznaczonego jako NM_000518.4 (NM oznacza  identyfikator  dla  mRNA,  a  NP  dla  sekwencji  białkowych).  Wyświetli  się podsumowanie tej sekwencji w formacie GenBank.
7. W prawym górnym rogu możesz zapisać sekwencję w formacie FASTA wybi erając Send (zaznacz opcję Coding Sequences).


### 2. 
Results found in 35 databases for "hemoglobin human"
### 3.
    - Summary
    
    The alpha (HBA) and beta (HBB) loci determine the structure of the 2 types of polypeptide chains in adult hemoglobin, Hb A. The normal adult hemoglobin tetramer consists of two alpha chains and two beta chains. Mutant beta globin causes sickle cell anemia. Absence of beta chain causes beta-zero-thalassemia. Reduced amounts of detectable beta globin causes beta-plus-thalassemia. The order of the genes in the beta-globin cluster is 5'-epsilon -- gamma-G -- gamma-A -- delta -- beta--3'. [provided by RefSeq, Jul 2008]
    
    - Genomic context
    - Genomic regions, transcripts, and products
    - Expression
    - Phenotypes
    - Variation
    - General gene information
    - General protein information
### 4. 
![title](res/2.1.png)
### 6.
![title](res/2.2.png)
![title](res/2.3.png)

## Zadanie 2. Operacje na sekwencjach.

In [1]:
from Bio.Seq import Seq
sequence = Seq("TGAGGCCCTGGGCAGGCTGCTGGTGGTCTACCCT")
print(sequence)
# obiekt Seq zachowuje się jak string:
print(sequence[0])
print(sequence[-1])
print(sequence.count('A'))
print(len(sequence))
print(sequence+'ATTTTTTA')

print(sequence[0::3]) # pierwszy nukleotyd z kodonu
print(sequence[1::3]) # drugi nukleotyd z kodonu
print(sequence[2::3]) # trzeci nukleotyd z kodonu

# mozemy przekształcić sekwencje do stringa:
print(str(sequence))

TGAGGCCCTGGGCAGGCTGCTGGTGGTCTACCCT
T
T
3
34
TGAGGCCCTGGGCAGGCTGCTGGTGGTCTACCCTATTTTTTA
TGCGCGGGGCCT
GGCGACCGGTC
ACTGGTTTTAC
TGAGGCCCTGGGCAGGCTGCTGGTGGTCTACCCT


## Zadanie 3. Alfabet
Obiekt  Seq,  reprezentujący  sekwencje,  zachowuje  się  tak  samo  jak  obiekt  typu  string, 
dodatk
owo  możemy  ustawić  w  nim  alfabet,  czyli  sprecyzować  czy  jest  to  sekwencja 
nukleotydowa czy może sekwencja złożona z aminokwasów. Nie można łączyć sekwencji o 
różnych alfabetach:

In [2]:
from Bio.Seq import Seq
from Bio.Alphabet import IUPAC
sequenceN=Seq("GATCGATGGGCCTATATAGGATCGAAAATCGC")
# alfabet obiektu
print(sequenceN.alphabet)
sequenceN = Seq("GATCGATGGGCCTATATAGGATCGAAAATCGC", IUPAC.unambiguous_dna)
print(sequenceN.alphabet)
sequenceAA = Seq("GATCGATGGGCCTATATAGGATCGAAAATCGC", IUPAC.protein)
print(sequenceN.alphabet)
print(sequenceAA.alphabet)
# nie można łączyć sekwencji o różnych alfabetach!
# print(sequenceN+sequenceAA)

Alphabet()
IUPACUnambiguousDNA()
IUPACUnambiguousDNA()
IUPACProtein()


## Zadanie 4. Nić komplementarna

In [3]:
from Bio.Seq import Seq
from Bio.Alphabet import IUPAC
nt_seq=Seq("ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG",IUPAC.unambiguous_dna)
print(nt_seq.complement()) 
#uwaga ta sekwencja jest w kierunku 3' do 5'
prot_seq=Seq("MVHLTPEEKSAVTALWGKVNVDEVGGEALGRLLVVYPWTQRFFE",IUPAC.protein)
# sekwencje białkowe nie mają komplementarnych
# print(prot_seq.complement()) 

TACCGGTAACATTACCCGGCGACTTTCCCACGGGCTATC


## Zadanie 5. Transkrypcja i translacja
Zwróć uwagę na automatyczną zmianę alfabetu

In [4]:
from Bio.Seq import Seq
from Bio.Alphabet import IUPAC
coding_dna=Seq("ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGA",IUPAC.unambiguous_dna)
print(coding_dna, coding_dna.alphabet)
#transkrypcja
messenger_rna = coding_dna.transcribe()
#translacja z mRNA
protein = messenger_rna.translate()
#translacja prosto z DNA
protein2 = coding_dna.translate()
print(messenger_rna, messenger_rna.alphabet)
print(protein, protein.alphabet)
print(protein2, protein2.alphabet)

ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGA IUPACUnambiguousDNA()
AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGA IUPACUnambiguousRNA()
MAIVMGR*KGAR HasStopCodon(IUPACProtein(), '*')
MAIVMGR*KGAR HasStopCodon(IUPACProtein(), '*')


## Zadanie 6. Obiekt SeqRecord

In [5]:
from Bio.Seq import Seq
from Bio.SeqRecord import SeqRecord
simple_seq = Seq("TGAGGCCCTGGGCAGGCTGCTGGTGGTCTACCC")
simple_seq_r = SeqRecord(simple_seq, id="AC12345",name="seqX",description="Homo sapiens, chr11")
print(simple_seq_r)

ID: AC12345
Name: seqX
Description: Homo sapiens, chr11
Number of features: 0
Seq('TGAGGCCCTGGGCAGGCTGCTGGTGGTCTACCC', Alphabet())


Do obiektu SeqRecord możemy bardzo łatwo wczytać plik typu FASTA

In [6]:
from Bio import SeqIO
record = SeqIO.read("data/hemoglobin1.fa", "fasta")
print(record.id)
print(record.name)
print(record.description)
print(record.annotations)
print(record.seq)

NG_000007.3:70545-72150
NG_000007.3:70545-72150
NG_000007.3:70545-72150 Homo sapiens beta globin region (HBB@); and beta globin locus transcript 3 (non-protein coding) (BGLT3); and hemoglobin subunit beta (HBB); and hemoglobin subunit delta (HBD); and hemoglobin subunit epsilon 1 (HBE1); and hemoglobin subunit gamma 1 (HBG1); and hemoglobin subunit gamma 2 (HBG2), RefSeqGene on chromosome 11
{}
ACATTTGCTTCTGACACAACTGTGTTCACTAGCAACCTCAAACAGACACCATGGTGCATCTGACTCCTGAGGAGAAGTCTGCCGTTACTGCCCTGTGGGGCAAGGTGAACGTGGATGAAGTTGGTGGTGAGGCCCTGGGCAGGTTGGTATCAAGGTTACAAGACAGGTTTAAGGAGACCAATAGAAACTGGGCATGTGGAGACAGAGAAGACTCTTGGGTTTCTGATAGGCACTGACTCTCTCTGCCTATTGGTCTATTTTCCCACCCTTAGGCTGCTGGTGGTCTACCCTTGGACCCAGAGGTTCTTTGAGTCCTTTGGGGATCTGTCCACTCCTGATGCTGTTATGGGCAACCCTAAGGTGAAGGCTCATGGCAAGAAAGTGCTCGGTGCCTTTAGTGATGGCCTGGCTCACCTGGACAACCTCAAGGGCACCTTTGCCACACTGAGTGAGCTGCACTGTGACAAGCTGCACGTGGATCCTGAGAACTTCAGGGTGAGTCTATGGGACGCTTGATGTTTTCTTTCCCCTTCTTTTCTATGGTTAAGTTCATGTCATAGGAAGGGGATAAGTAACAGGGTACAGTTTAGAATGGGAAACAG

## Zadanie 7. Zobacz jakie bazy danych są dostępne przez Bio.Entrez

In [7]:
from Bio import Entrez
Entrez.email = "your_email@uj.edu.pl"
# przedstawiamy sie
handle = Entrez.einfo()
result = handle.read()
print(result)
# otrzymujemy XML
handle = Entrez.einfo()
result = Entrez.read(handle) 
# uzywajac parsera z Bio.Entrez
print(result)
# otrzymujemy obiekt Pythona - slownik z lista baz

<?xml version="1.0" encoding="UTF-8" ?>
<!DOCTYPE eInfoResult PUBLIC "-//NLM//DTD einfo 20130322//EN" "https://eutils.ncbi.nlm.nih.gov/eutils/dtd/20130322/einfo.dtd">
<eInfoResult>
<DbList>

	<DbName>pubmed</DbName>
	<DbName>protein</DbName>
	<DbName>nuccore</DbName>
	<DbName>ipg</DbName>
	<DbName>nucleotide</DbName>
	<DbName>nucgss</DbName>
	<DbName>nucest</DbName>
	<DbName>structure</DbName>
	<DbName>sparcle</DbName>
	<DbName>genome</DbName>
	<DbName>annotinfo</DbName>
	<DbName>assembly</DbName>
	<DbName>bioproject</DbName>
	<DbName>biosample</DbName>
	<DbName>blastdbinfo</DbName>
	<DbName>books</DbName>
	<DbName>cdd</DbName>
	<DbName>clinvar</DbName>
	<DbName>clone</DbName>
	<DbName>gap</DbName>
	<DbName>gapplus</DbName>
	<DbName>grasp</DbName>
	<DbName>dbvar</DbName>
	<DbName>gene</DbName>
	<DbName>gds</DbName>
	<DbName>geoprofiles</DbName>
	<DbName>homologene</DbName>
	<DbName>medgen</DbName>
	<DbName>mesh</DbName>
	<DbName>ncbisearch</DbName>
	<DbName>nlmcatalog</DbName>
	<DbName

## Zadanie 8. Zobacz informacje o konkretnej bazie

In [8]:
handle = Entrez.einfo(db="nucleotide")
result = Entrez.read(handle)
print(result["DbInfo"].keys())
print(result["DbInfo"]["Description"])
print(result["DbInfo"]["Count"])
print(result["DbInfo"]["LastUpdate"])
for field in result["DbInfo"]["FieldList"]: 
    print("%(Name)s, %(FullName)s, %(Description)s" % field)

dict_keys(['DbName', 'MenuName', 'Description', 'DbBuild', 'Count', 'LastUpdate', 'FieldList', 'LinkList'])
Core Nucleotide db
242043317
2017/07/28 12:38
ALL, All Fields, All terms from all searchable fields
UID, UID, Unique number assigned to each sequence
FILT, Filter, Limits the records
WORD, Text Word, Free text associated with record
TITL, Title, Words in definition line
KYWD, Keyword, Nonstandardized terms provided by submitter
AUTH, Author, Author(s) of publication
JOUR, Journal, Journal abbreviation of publication
VOL, Volume, Volume number of publication
ISS, Issue, Issue number of publication
PAGE, Page Number, Page number(s) of publication
ORGN, Organism, Scientific and common names of organism, and all higher levels of taxonomy
ACCN, Accession, Accession number of sequence
PACC, Primary Accession, Does not include retired secondary accessions
GENE, Gene Name, Name of gene associated with sequence
PROT, Protein Name, Name of protein associated with sequence
ECNO, EC/RN Numbe

Sprawdź informacje dla innych baz np. protein, structure bądź pubmed.

In [9]:
handle = Entrez.einfo(db="protein")
result = Entrez.read(handle)
print(result["DbInfo"].keys())
print(result["DbInfo"]["Description"])
print(result["DbInfo"]["Count"])
print(result["DbInfo"]["LastUpdate"])
for field in result["DbInfo"]["FieldList"]: 
    print("%(Name)s, %(FullName)s, %(Description)s" % field)

dict_keys(['DbName', 'MenuName', 'Description', 'DbBuild', 'Count', 'LastUpdate', 'FieldList', 'LinkList'])
Protein sequence record
412823586
2017/07/24 19:00
ALL, All Fields, All terms from all searchable fields
UID, UID, Unique number assigned to each sequence
FILT, Filter, Limits the records
WORD, Text Word, Free text associated with record
TITL, Title, Words in definition line
KYWD, Keyword, Nonstandardized terms provided by submitter
AUTH, Author, Author(s) of publication
JOUR, Journal, Journal abbreviation of publication
VOL, Volume, Volume number of publication
ISS, Issue, Issue number of publication
PAGE, Page Number, Page number(s) of publication
ORGN, Organism, Scientific and common names of organism, and all higher levels of taxonomy
ACCN, Accession, Accession number of sequence
PACC, Primary Accession, Does not include retired secondary accessions
GENE, Gene Name, Name of gene associated with sequence
PROT, Protein Name, Name of protein associated with sequence
ECNO, EC/RN 

## Zadanie 9. Wyszukaj rekordy z bazy nucleotide, które posiadają związek z wirusem świńskiej grypy

In [10]:
handle = Entrez.esearch(db="nucleotide",term="swine influenza",retmax="10")
result = Entrez.read(handle)
print(result["IdList"])

['1219400587', '1219400585', '1219400583', '1219400580', '1219400577', '1219400575', '1219400573', '1219400571', '1219400568', '1219400566']


Parametr „retmax” określa maksymalną liczbę zwracanych rezultatów. 
Wyszukaj te same rekordy przez przegladarkę, zobacz jakiemu polu rekordu odpowiadają unikalne numery sekwencji nukleotydowych, zwrócone 
w tym zapytaniu.
Używając  AND,  OR  i  NOT  oraz  pól  z  FieldList  znalezionych  wcześniej  przez EInfo, możemy rozbudować zapytanie, np.:

In [11]:
term = "influenza AND H1N1[TITL] AND (hemagglutinin[PROT] OR neuraminidase[PROT])"

wyszuka  wszystkie  rekordy  czyli  sekwencje  nukleotydowe,  które  w  którymkolwiek  polu zawierają słowo „influenza”, zawierają w tytule „H1N1” i kodują białka hemagglutinin bądź 
neuraminidase (są to dwa bardzo ważne białka obecne na powierzchni wirusa, są one często celem leków przeciwwirusowych, od
ich typów wirusy grypy biorą swoje nazwy - np. H1N1)

Wymyśl  i  wypróbuj  jakieś  swoje  zapytanie,  używając  spójników  i  pól  wyszukiwania  z 
FieldList
(może być dla innej bazy niż nucleotide).

In [12]:
myterm = "Avian influenza AND (H2N2[TITL] OR glycoprotein[PROT])"
handle = Entrez.esearch(db="nucleotide", term = myterm, retmax="10")
result = Entrez.read(handle)
print(result["IdList"])

['8927539', '8927537', '8927535', '8927533', '8927531', '8927529', '8927527', '8927525', '8927523', '8927521']


## Zadanie 10. Za pomocą ESummary pobierz informację podsumowującą o rekordzie

Zmodyfikuj poprzedni przykład używając:

    handle  =  Entrez.esummary(db="nucleotide",  id="326579398") # pojedynczy rekord

Można też pobrać info o wielu rekordach, podając na wejściu listę identyfikatorów.

    handle = Entrez.esummary(db="nucleotide", id=",".join(result["IdList"])) #lista rekordow

Otrzymujemy  wówczas  listę  słowników – w  każdym  słowniku  jest  informacja  o  jednym rekordzie, którą można wypisać np. tak:

    result = Entrez.read(handle)
    for record in result:
        print("%(Title)s, %(Gi)s, %(Length)s %(CreateDate)s" % record)

In [13]:
handle = Entrez.esummary(db="nucleotide", id=",".join(result["IdList"])) #lista rekordow
result = Entrez.read(handle)
for record in result:
    print("%(Title)s, %(Gi)s, %(Length)s %(CreateDate)s" % record)

Influenza A virus isolate A/Chile/6/57 (H2N2)) hemagglutinin gene, partial cds, 8927539, 1017 2000/07/06
Influenza A virus isolate A/Ohio/2/59 (H2N2)) hemagglutinin gene, partial cds, 8927537, 1017 2000/07/06
Influenza A virus isolate A/Victoria/15681/59 (H2N2)) hemagglutinin gene, partial cds, 8927535, 1017 2000/07/06
Influenza A virus isolate A/Sao Paolo/3/59 (H2N2)) hemagglutinin gene, partial cds, 8927533, 1017 2000/07/06
Influenza A virus isolate A/Malaya/16/58 (H2N2)) hemagglutinin gene, partial cds, 8927531, 1017 2000/07/06
Influenza A virus isolate A/Albany/6/58 (H2N2)) hemagglutinin gene, partial cds, 8927529, 1017 2000/07/06
Influenza A virus isolate A/RI/5+/57 (H2N2)) hemagglutinin gene, partial cds, 8927527, 1017 2000/07/06
Influenza A virus isolate A/Ann Arbor/6/60 (H2N2)) hemagglutinin gene, partial cds, 8927525, 1017 2000/07/06
Influenza A virus isolate A/Albany/7/57 (H2N2)) hemagglutinin gene, partial cds, 8927523, 1017 2000/07/06
Influenza A virus isolate A/Davis/1/57 

## Zadanie 11. Pobierz cały rekord

Aby pobrać cały rekord korzystamy z EFetch, możemy też wybrać inny format niż XML, np. podstawowe formaty jak genbank, fasta, genpept dla bazy Protein (dokładny opis formatów:
www.ncbi.nlm.nih.gov/entrez/query/static/efetchseq_help.html)

In [14]:
handle = Entrez.efetch(db="nucleotide", id="326579398", rettype="gb", retmode="text")
print(handle.read())

LOCUS       JF703685                2177 bp    cRNA    linear   VRL 30-NOV-2011
DEFINITION  Influenza A virus (A/swine/Quebec/1265553/2010(H3N2)) segment 3
            polymerase PA (PA) gene, complete cds.
ACCESSION   JF703685
VERSION     JF703685.1
KEYWORDS    .
SOURCE      Influenza A virus (A/swine/Quebec/1265553/2010(H3N2))
  ORGANISM  Influenza A virus (A/swine/Quebec/1265553/2010(H3N2))
            Viruses; ssRNA viruses; ssRNA negative-strand viruses;
            Orthomyxoviridae; Influenzavirus A.
REFERENCE   1  (bases 1 to 2177)
  AUTHORS   Tremblay,D., Allard,V., Doyon,J.F., Bellehumeur,C., Spearman,J.G.,
            Harel,J. and Gagnon,C.A.
  TITLE     Emergence of a New Swine H3N2 and Pandemic (H1N1) 2009 Influenza A
            Virus Reassortant in Two Canadian Animal Populations, Mink and
            Swine
  JOURNAL   J. Clin. Microbiol. 49 (12), 4386-4390 (2011)
   PUBMED   22012020
REFERENCE   2  (bases 1 to 2177)
  AUTHORS   Tremblay,D., Allard,V., Harel,J. and Gagnon

Plik w odpowiednim formacie możemy od razu wczytać do obiektu sekwencji SeqRecord:

In [15]:
from Bio import SeqIO
handle = Entrez.efetch(db="nucleotide", id="326579398", rettype="gb", retmode="text")
record = SeqIO.read(handle, "genbank")
print(record.name)
print(record.description)
print(record.seq)

JF703685
Influenza A virus (A/swine/Quebec/1265553/2010(H3N2)) segment 3 polymerase PA (PA) gene, complete cds
CAAAATGGAAGACTTTGTGCGACAATGCTTCAATCCAATGATCGTCGAGCTTGCGGAAAAGGCAATGAAAGAATATGGGGAAGATCCGAAAATCGAAACTAACAAGTTTGCTGCAATATGCACACATTTGGAAGTTTGTTTCATGTATTCGGATTTCCATTTCATCGACGAACGAGGTGAATCAATAATTGTAGAATCTGGTGACCCGAATGCACTATTGAAGCACCGATTTGAGATAATTGAAGGAAGAGACCGAATCATGGCCTGGACAGTGGTGAACAGTATATGTAACACAACAGGGGTAGAGAAGCCTAAATTTCTTCCTGATTTGTATGATTACAAAGAGAACCGGTTCATTGAAATTGGAGTAACACGGAGGGAAGTCCACATATATTACCTAGAGAAAGCCAACAAAATAAAATCTGAGAAGACACACATTCACATCTTTTCATTCACTGGAGAGGAGATGGCCACCAAAGCGGACTACACCCTTGACGAAGAGAGCAGGGCAAGAATCAAAACTAGGCTTTTCACTATAAGACAAGAAATGGCCAGTAGGAGTCTATGGGATTCCTTTCGTCAGTCCGAAAGAGGCGAAGAGACAATTGAAGAAAAATTTGAGATTAAAGGAACTATGCGCAAGCTTGCCGACCAAAGTCTCCCACCGAACTTCTCCAGCCTTGAAAACTTTAGAGCCTATGTAGATGGATTCGAGCCGAACGGCTGCATTGAGGGCAAGCTTTCCCAAATGTCAAAAGAAGTGAACGCCAAAATTGAACCATTCTTGAGGACGACACCACGCCCCCTCAGATTGCCTGGTGGGCCTCTTTGCCATCAGCGGTCAAAGTTCCTGCTGATGGATGCTCTGAAATTAAGTATTGAAGACCCG

## Zadanie 12. Historia sesji
NCBI  zapewnia  możliwość  korzystania  z  historii  sesji,  stąd  można  usprawnić  proces wyszukiwania  i
ściągania  sekwencji  za  pomocą ESearch i EFetch,  tak  by  robić  do  bez podawania listy uzyskanych ID, wykonując 
ESearch użyjmy opcji usehistory,  dostaniemy dwie nowe pozycje - klucz zapytania i ciasteczko sesji:

In [16]:
handle = Entrez.esearch(db="protein", term="Homo sapiens[orgn] and rhodopsin", usehistory="y")
result = Entrez.read(handle)
querykey = result['QueryKey']
webenv = result['WebEnv']
print(querykey)
print(webenv)

1
NCID_1_31888426_130.14.22.215_9001_1501289902_1075383314_0MetA0_S_MegaStore_F_1


Możemy  ich  teraz  użyć  bezpośrednio  w EFetch, ponieważ  z  reguły  wyniki  takiego wyszukiwania będą bardzo duże, możemy wybrać zakres rekordów, który chcemy obejrzeć 
przez retstart i retmax.

In [17]:
handle = Entrez.efetch(db="protein", WebEnv=webenv, query_key=querykey, retstart=0, retmax=2, rettype="gp", retmode="text")
print(handle.read())

LOCUS       NP_005293                613 aa            linear   PRI 19-JUL-2017
DEFINITION  prosaposin receptor GPR37 precursor [Homo sapiens].
ACCESSION   NP_005293
VERSION     NP_005293.1
DBSOURCE    REFSEQ: accession NM_005302.4
KEYWORDS    RefSeq.
SOURCE      Homo sapiens (human)
  ORGANISM  Homo sapiens
            Eukaryota; Metazoa; Chordata; Craniata; Vertebrata; Euteleostomi;
            Mammalia; Eutheria; Euarchontoglires; Primates; Haplorrhini;
            Catarrhini; Hominidae; Homo.
REFERENCE   1  (residues 1 to 613)
  AUTHORS   Berger BS, Acebron SP, Herbst J, Koch S and Niehrs C.
  TITLE     Parkinson's disease-associated receptor GPR37 is an ER chaperone
            for LRP6
  JOURNAL   EMBO Rep. 18 (5), 712-725 (2017)
   PUBMED   28341812
REFERENCE   2  (residues 1 to 613)
  AUTHORS   Tanabe Y, Fujita-Jimbo E, Momoi MY and Momoi T.
  TITLE     CASPR2 forms a complex with GPR37 via MUPP1 but not with
            GPR37(R558Q), an autism spectrum disorder-related mutatio

## Zadanie 13. Elink
Elink pozwala na wyszukanie linków pomiędzy bazami, np. mając ID genu chcemy przejść do sekwencji białkowej

In [18]:
handle = Entrez.elink(dbfrom="gene", db="protein", id=3043) 
# ID genu hemoglobiny czlowieka
result = Entrez.read(handle)
for link in result[0]["LinkSetDb"][0]["Link"]:
    print(link["Id"])  
#wypisze listę powiązanych identyfikatorów z bazy protein

1043914264
1005667216
970597826
970597824
918400562
890880780
772394756
772394754
772394752
772394750
772394748
772394746
772394744
772394742
772394740
772394729
772394727
772394725
772394723
772394721
772394719
772394717
772394715
772394713
772394711
544336747
410178713
410178711
402830039
350285456
321271362
319739575
270297796
270297794
256028940
237675828
223971865
223970113
194719364
193244963
193244961
193244959
193244957
193244955
193244953
193244951
193244949
193244947
193244945
193244943
193244941
193244939
193244937
193244935
193244933
193244931
193244929
193244927
193244925
193244923
193244921
193244919
193244917
193244915
193244913
193244911
193244909
193244907
193244905
193244903
193244901
193244899
193244897
193244895
193244893
193244891
193244889
193244887
193244885
193244883
193244881
193244879
193244877
193244875
193244873
193244871
193244869
193244867
190613770
189053145
187940241
186461228
184186850
184186848
163140875
157928558
133711987
123999887
123992967
11958921

W pętli for pierwsze [0] dlatego, że szukamy linków dla pierwszego id (w naszym przypadku jest tylko jeden, ale parametrem elink może być również lista identyfikatorów). Drugie [0] dlatego, że szukamy dla pierwszej bazy docelowej (w naszym przypadku „protein”; uwaga j.w. – tu  jest  jedna  baza  docelowa,  ale  może  być więcej).  Dodatkowe  informacje  o parametrach Elinka znajdziesz na stronie NCBI Eutils.

## Zadanie 14.
Napisz skrypt, który przy wykonuje kolejno następujące zadania:
    1. Wyszukaj geny powiązane z nazwą BRCA1 u człowieka. Wypisz pola: Name, Chromosome, MapLocation i Description


In [19]:
item = 'BRCA1'
animal = 'Homo sapiens' 
search_string = item + " AND " + animal + "[Organism]"

handle = Entrez.esearch(db="gene", term=search_string)
resultSearch = Entrez.read(handle)
handle = Entrez.esummary(db="gene", id=",".join(resultSearch["IdList"]))
result = Entrez.read(handle)

for record in result['DocumentSummarySet']['DocumentSummary']:
    print("%(Name)s, %(Chromosome)s, %(MapLocation)s, %(Description)s" % record)    

TP53, 17, 17p13.1, tumor protein p53
EGFR, 7, 7p11.2, epidermal growth factor receptor
VEGFA, 6, 6p21.1, vascular endothelial growth factor A
MTHFR, 1, 1p36.22, methylenetetrahydrofolate reductase
TGFB1, 19, 19q13.2, transforming growth factor beta 1
ESR1, 6, 6q25.1-q25.2, estrogen receptor 1
ERBB2, 17, 17q12, erb-b2 receptor tyrosine kinase 2
HIF1A, 14, 14q23.2, hypoxia inducible factor 1 alpha subunit
BRCA1, 17, 17q21.31, BRCA1, DNA repair associated
ABCB1, 7, 7q21.12, ATP binding cassette subfamily B member 1
NFKB1, 4, 4q24, nuclear factor kappa B subunit 1
AR, X, Xq12, androgen receptor
AKT1, 14, 14q32.33, AKT serine/threonine kinase 1
CDKN2A, 9, 9p21.3, cyclin dependent kinase inhibitor 2A
IL1B, 2, 2q14.1, interleukin 1 beta
VDR, 12, 12q13.11, vitamin D receptor
KRAS, 12, 12p12.1, KRAS proto-oncogene, GTPase
STAT3, 17, 17q21.2, signal transducer and activator of transcription 3
COMT, 22, 22q11.21, catechol-O-methyltransferase
PTEN, 10, 10q23.31, phosphatase and tensin homolog


    2. Zachowaj id genu o nazwie BRCA1 

In [20]:
handle = Entrez.esearch(db="gene", term=item)
search = Entrez.read(handle)
handle = Entrez.esummary(db="gene", id=",".join(search["IdList"]))
result = Entrez.read(handle)

for i, record in enumerate (result['DocumentSummarySet']['DocumentSummary']):
    if record['Name'] == "BRCA1":
        BRCA1_Id = search["IdList"][i]

print(BRCA1_Id)

672


    3. Wypisz tytuły rekordów bazy OMIM związane z tym genem


In [21]:
handle = Entrez.esearch(db="OMIM", term=item)
search = Entrez.read(handle)
for i in search["IdList"]:
        handle = Entrez.esummary(db="OMIM", id=i)
        result = Entrez.read(handle)
        print(result[0]["Title"])

ZINC FINGER PROTEIN 589; ZNF589
UBX DOMAIN PROTEIN 1; UBXN1
GLUTATHIONE PEROXIDASE 7; GPX7
CYCLIN-DEPENDENT KINASE 12; CDK12
BRCA1-ASSOCIATED ATM ACTIVATOR 1; BRAT1
RIGIDITY AND MULTIFOCAL SEIZURE SYNDROME, LETHAL NEONATAL; RMFSL
TUMOR PREDISPOSITION SYNDROME; TPDS
PANCREATIC CANCER, SUSCEPTIBILITY TO, 4; PNCA4
FANCD2 GENE; FANCD2
FANCE GENE; FANCE
FANCC GENE; FANCC
GLUCOSE-6-PHOSPHATASE, CATALYTIC; G6PC
MEN1 GENE; MEN1
BREAST-OVARIAN CANCER, FAMILIAL, SUSCEPTIBILITY TO, 3; BROVCA3
PANCREATIC CANCER, SUSCEPTIBILITY TO, 3
PANCREATIC CANCER, SUSCEPTIBILITY TO, 2
NEUROFIBROMIN 1; NF1
CHROMOSOME 19 OPEN READING FRAME 62; C19ORF62
RING FINGER PROTEIN 168; RNF168
BREAST-OVARIAN CANCER, FAMILIAL, SUSCEPTIBILITY TO, 2; BROVCA2


    4. Znajdź białka powiązane z tym genem (baza protein): wypisz pola numeru identyfikacyjnego Gi dla wyszukanych sekwencji


In [22]:
handle = Entrez.esearch(db="protein", term=search_string)
search = Entrez.read(handle)
for i in search["IdList"]:
        handle = Entrez.esummary(db="protein", id=i)
        result = Entrez.read(handle)
        print(result[0]["Gi"])

1007353152
1007352673
1007352648
929524271
929524269
929524267
862669351
336285487
330864753
217416415
217272816
190684675
156523968
148529023
112789562
67189945
64762484
33356172
19924125
19923717


    5. Pobierz białko o numerze identyfikacyjnym 121949022 w formacie genbank, wypisz jego nazwę oraz sekwencję aminokwasów


In [23]:
handle = Entrez.efetch(db="protein", id="121949022", rettype="gb", retmode="text")
result = SeqIO.read(handle, "genbank")
print(result.name)
print(result.seq)

Q3LRJ6_HUMAN
MDLSALRVEEVQNVINAMQKILECPICLELIKEPVSTKCDHIFCKFCMLKLLNQKKGPSQCPLCKNDITKRSLQESTRFSQLVEELLKIICAFQLDTGLEYANSYNFAKKENNSPEHLKDEVSIIQSMGYRNRAKRLLQSEPENPSLQETSLSVQLSNLGTVRTLRTKQRIQPQKTSVYIELGSDSSEDTVNKATYCSVGDQELLQITPQGTRDEISLDSAKKAACEFSETDVTNTEHHQPSNNDLNTTEKRAAERHPEKYQGSSVSNLHVEPCGTNTHASSLQHENSSLLLTKDRMNVEKAEFCNKSKQPGLARSQHNRWAGSKETCNDRRTPSTEKKVDLNADPLCERKEWNKQKLPCSENPRDTEDVPWITLNSSIQKVNEWFSRSDELLGSDDSHDGESESNAKVADVLDVLNEVDEYSGSSEKIDLLASDPHEALICKSERVHSKSVESNIEDKIFGKTYRKKASLPNLSHVTENLIIGAFVTEPQIIQERPLTNKLKRKRRPTSGLHPEDFIKKADLAVQKTPEMINQGTNQTEQNGQVMNITNSGHENKTKGDSIQNEKNPNPIESLEKESAFKTKAEPISSSISNMELELNIHNSKAPKKNRLRRKSSTRHIHALELVVSRNLSPPNCTELQIDSCSSSEEIKKKKYNQMPVRHSRNLQLMEGKEPATGAKKSNKPNEQTSKRHDSDTFPELKLTNAPGSFTKCSNTSELKEFVNPSLPREEKEEKLETVKVSNNAEDPKDLMLSGERVLQTERSVESSSISLVPGTDYGTQESISLLEVSTLGKAKTEPNKCVSQCAAFENPKGLIHGCSKDNRNDTEGFKYPLGHEVNHSRETSIEMEESELDAQYLQNTFKVSKRQSFAPFSNPGNAEEECATFSAHSGSLKKQSPKVTFECEQKEENQGKNESNIKPVQTVNITAGFPVVGQKDKPVDNAKCSIKGGSRFCLSSQFRGNETGLITPNKHGLLQNPYRIPPLFPIK

    6. Znajdź wszystkie mutacje w obrębie genu BRCA1 u człowieka (w elink do bazy SNP użyj term="Homo sapiens"), wypisz ile mutacji znaleziono, dla pierwszych 50 wyświetl ich numer identyfikacyjny SNP_ID, klasę SNP_CLASS, gen GENE i pozycję CONTIGPOS


In [24]:
handleLink = Entrez.elink(dbfrom="gene", db="snp", id=BRCA1_Id, term=animal)
resultLink = Entrez.read(handleLink )

linkList = []
for link in resultLink[0]['LinkSetDb'][0]['Link']:
    linkList.append(link["Id"])

print("Count of mutatons:", len(linkList))

handle = Entrez.esummary(db="SNP",id=",".join(linkList[:50]))
result = Entrez.read(handle)

for record in result:
    print("%(SNP_ID)s %(SNP_CLASS)s %(GENE)s %(CONTIGPOS)s" % record)

Count of mutatons: 7141
864622737 snp  NT_010783.16:16168869
864622723 snp  NT_010783.16:16158416
864622618 snp  NT_010783.16:16156710
864622588 snp  NT_010783.16:16156804
864622540 snp  NT_010783.16:16154967
864622536 in-del  NT_010783.16:16157172
864622534 snp  NT_010783.16:16179817
864622491 snp  NT_010783.16:16156947
864622454 snp  NT_010783.16:16158257
864622452 snp  NT_010783.16:16157553
864622444 snp  NT_010783.16:16168902
864622350 in-del  NT_010783.16:16163810
864622306 snp  NT_010783.16:16159960
864622265 snp  NT_010783.16:16138385
864622260 snp  NT_010783.16:16163877
864622244 snp  NT_010783.16:16113175
864622233 snp  NT_010783.16:16155620
864622220 in-del  NT_010783.16:16109765
864622146 snp  NT_010783.16:16156083
864622132 in-del  NT_010783.16:16135224
864622124 snp  NT_010783.16:16168201
864622122 snp  NT_010783.16:16156984
864622119 in-del  NT_010783.16:16146418
864622104 snp  NT_010783.16:16121124
864622080 snp  NT_010783.16:16155978
863224841 in-del  NT_010783.16:16157

    7. Na podstawie przeprowadzonych wyżej poszukiwań, sformułuj odpowiednie wnioski (jakie informacje udało Ci się uzyskać? jakie jest ich znaczenie?). Treść wniosku wyświetl na ekranie (print) lub zawrzyj w komentarzu na końcu pliku ze skryptem.


1. BRCA1 (breast cancer 1) "DNA repair associated", help repair damaged DNA, or destroy cells if DNA cannot be repaired. They are involved in the repair of chromosomal damage with an important role in the error-free repair of DNA double-strand breaks. Many genes responsible for other functions are associated with it, for example, receptors and responsible for growth.

2. Its id is 672, which can be used for more detailed search in databases.

3. OMIM is a continuously updated catalog of human genes and genetic disorders and traits, with a particular focus on the gene-phenotype relationship. A large number of records showing that gene mutations can lead to different types of cancer. If BRCA1 itself is damaged by a BRCA mutation, damaged DNA is not repaired properly, and this increases the risk for breast cancer.

4. (5) We can find proteins related request, as well as get their name and sequence using the database.

6. There are many other mutations by single nucleotide polymorphism (when there is single base which differs between two genomes, and the DNA around that base is otherwise unchanged) or INsertion/DELetion (where a single base has been deleted, or inserted into one genome relative to another) of this gene and some of them are quite dangerous.